suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/gates/SpdzMacCheckProtocol.java

```diff
@@ -36,95 +38,80 @@ public SpdzMacCheckProtocol(
       final MessageDigest digest,
       final SpdzStorage storage,
       final BigInteger modulus) {
-    protocolProducer = new SequentialProtocolProducer();
-    protocolProducer.append(new LazyProtocolProducerDecorator(() -> {
-      BigInteger s = new BigInteger(modulus.bitLength(), rand).mod(modulus);
-      SpdzCommitment commitment = new SpdzCommitment(digest, s, rand);
-      SpdzCommitProtocol commitmentProtocol = new SpdzCommitProtocol(commitment);
-      openComm = new SpdzOpenCommitProtocol(commitment, commitmentProtocol);
-
-      return new SequentialProtocolProducer(
-          new SingleProtocolProducer<>(commitmentProtocol),
-          new SingleProtocolProducer<>(openComm));
-    }));
-    protocolProducer.append(new LazyProtocolProducerDecorator(() -> {
-      List<BigInteger> openedValues = storage.getOpenedValues();
-
-      // Add all s's to get the common random value:
-      BigInteger s = BigInteger.ZERO;
-      Map<Integer, BigInteger> commitments = openComm.out();
-      for (BigInteger otherS : commitments.values()) {
-        s = s.add(otherS);
-      }
+    this.rand = rand;
+    this.digest = digest;
+    this.storage = storage;
+    this.modulus = modulus;
+  }
 
-      openedValuesSize = openedValues.size();
+  @Override
+  public DRes<Void> buildComputation(ProtocolBuilderNumeric builder) {
+    SpdzCommitment commitment = new SpdzCommitment(digest,
+        new BigInteger(modulus.bitLength(), rand).mod(modulus), rand);
+    return builder.seq((seq) -> seq.append(new SpdzCommitProtocol(commitment)))
+        .seq((seq, commitProtocol) ->
+            seq.append(new SpdzOpenCommitProtocol(commitment, commitProtocol)))
+        .seq((seq, openCommit) -> {
+          List<BigInteger> openedValues = storage.getOpenedValues();
 
-      BigInteger[] rs = new BigInteger[openedValuesSize];
-      BigInteger temporaryR = s;
-      for (int i = 0; i < openedValuesSize; i++) {
-        temporaryR = new BigInteger(digest.digest(temporaryR.toByteArray())).mod(modulus);
-        rs[i] = temporaryR;
-      }
-      BigInteger a = BigInteger.ZERO;
-      int index = 0;
-      for (BigInteger aa : openedValues) {
-        a = a.add(aa.multiply(rs[index++])).mod(modulus);
-      }
+          // Add all s's to get the common random value:
+          BigInteger sum = BigInteger.ZERO;
+          for (BigInteger otherS : openCommit.values()) {
```

Comment:
How old fashioned ... Why not `openCommit.stream.reduce(ZERO, BigInteger::add);`